keps/sig-api-machinery/20200123-client-go-ctx.md

```diff
@@ -0,0 +1,188 @@
+- [Design Details](#design-details)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Graduation Criteria](#graduation-criteria)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This proposal outlines a refactoring that would allow propagating of
+[context](https://golang.org/pkg/context/) with
+[client-go](https://github.com/kubernetes/client-go) with API calls made using
+k8s.io/client-go clientsets.
+
+## Motivation
+
+When using client-go, external calls to the Kubernetes API are made in order to
+manage resources. Adding support for context propagation along these request
+paths enables clean and efficient implementation of cross cutting functionality.
+This functionality initially will include:
+
+- Support for request timeout and cancelation that frees the calling thread. Due
+  to issues with authentication and authorization performance, this was
+  something that was
+  [retrofited](https://github.com/kubernetes/kubernetes/pull/83064) in various
+  auth\* expansions but is generally useful.
+- Support for distributed tracing. Active
+  [efforts](https://github.com/kubernetes/enhancements/blob/master/keps/sig-instrumentation/0034-distributed-tracing-kep.md#propagating-context-through-objects)
+  by SIG Instrumentation to support distributed tracing are either blocked or
+  would be substantially simplified with this enhancement.
+
+### Goals
+
+- Allow consumers of k8s.io/client-go to propage context with API requests made
+  using a clientset.
+
+### Non-Goals
+
+- Plumb context to client-go callsites in core Kubernetes. We believe that by
+  making this feature available, this type of refactoring will happen over time.
+
+## Proposal
+
+The recommended approach is to pass the context with each action on a resource.
+We will modify the signatures of all client interfaces to accept a
+`context.Context` as the first argument.  This is idiomatic, explicit, and
+produces the least error prone API.
+
+## Design Details
+
+This is an example change to `Pods.Get` implementation:
+
+```go 
+func (c *pods) Get(ctx context.Context, name string, options meta_v1.GetOptions) (result *v1.Pod, err error) {
+	Result = &v1.Pod{}
+	err = c.client.Get().
+		Context(ctx).
+		Namespace(c.ns).
+		Resource("pods").
+		Name(name).
+		VersionedParams(&options, scheme.ParameterCodec).
+		Do().
+		Into(result)
+	return
+}
+```
+
+Here we extend the signature to add `context.Context` as the first argument,
+then attach the context to the request.
+
+Initially `lister` and `informer` APIs (both generated and generic) will not
+change.
+
+### Risks and Mitigations
+
+This is a breaking change that would likely require extensive changes to both
+tests and implementation code throughout the impacted code bases when client-go
+is upgraded.
+
+Inside the Kubernetes codebase, we will refactor all callsites to pass
+`context.TODO()` in the initial pass.
+
+TODO(mikedanese): Discuss and document migration path for external codebases
```

Comment:
@deads2k can you explain what you had in mind here?

Reply:
For each of the generated clientsets we have (`k8s.io/{client-go,kube-aggregator,apiextensions-apiserver}`), snapshot the existing generated clientsets under a peer "deprecated" package. For example, for `k8s.io/client-go/kubernetes`:

```sh
cp -r \
  staging/src/k8s.io/client-go/kubernetes \
  staging/src/k8s.io/client-go/deprecated

find staging/src/k8s.io/client-go/deprecated -type f -print0 | 
  xargs -n 1 -0 sed -i 's#k8s\.io/client-go/kubernetes#k8s.io/client-go/deprecated#g'
```

External codebases can then do an equivalent find/replace to use the deprecated clientset packages, and spend a release migrating to the context-aware clientsets.

Reply:
What's the benefit to this over not upgrading external codebase's client-go?

Reply:
sed of imports is easier than a go refactoring tool? I think the benefit is slight, so I'll let @deads2k defend it, but if it makes a difference between getting context threading or not, I'll live with the deprecated packages for a release

Reply:
Jordan described it well. This is the method that @lavalamp independently described.  The sed is easy and non-conflicty across a large codebase in contrast to the conflicts approach of changing every client call at once and before anyone can add another usage.  It allows phasing the change in after updating clients.

Reply:
How do you feel about `k8s.io/{metrics,node-api}/pkg/client/clientset`? I'll throw those in with `k8s.io/{client-go,kube-aggregator,apiextensions-apiserver}` until I hear otherwise and won't bother with the example clientsets.

Reply:
`k8s.io/node-api` is getting dropped in https://github.com/kubernetes/kubernetes/pull/87503

I'd forgotten about k8s.io/metrics generated clientsets... PodMetricsesGetter... sigh. I guess.

Reply:
Ok, I paraphrased the approach and motivation.